pkg/s3/s3client.go

```diff
@@ -0,0 +1,167 @@
+	}
+
+	pathStyle := true
+	token := ""
+	creds := credentials.NewStaticCredentials(accessKey, secret, token)
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config: aws.Config{
+			Region:           &region,
+			Credentials:      creds,
+			Endpoint:         &endpoint,
+			DisableSSL:       &noSSL,
+			S3ForcePathStyle: &pathStyle,
+		},
+		// Profile: "profile_name",
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create S3 session instance: %v", err)
+	}
+
+	return &S3Client{
+		S3: s3.New(sess),
+	}, nil
+}
+
+// CreateBucketIfNotExists creates the s3 bucket with name <bucket> in <region>.
+func (c *S3Client) CreateBucketIfNotExists(ctx context.Context, bucket, region string) error {
+	createBucketInput := &s3.CreateBucketInput{
+		Bucket: aws.String(bucket),
+		ACL:    aws.String(s3.BucketCannedACLPrivate),
+		CreateBucketConfiguration: &s3.CreateBucketConfiguration{
+			LocationConstraint: aws.String(region),
+		},
+	}
+
+	if _, err := c.S3.CreateBucketWithContext(ctx, createBucketInput); err != nil {
+		if aerr, ok := err.(awserr.Error); !ok {
+			return err
+		} else if aerr.Code() != s3.ErrCodeBucketAlreadyExists && aerr.Code() != s3.ErrCodeBucketAlreadyOwnedByYou {
+			return err
+		}
+	}
+	// Set lifecycle rule to purge incomplete multipart upload orphaned because of force shutdown or rescheduling or networking issue with etcd-backup-restore.
+	putBucketLifecycleConfigurationInput := &s3.PutBucketLifecycleConfigurationInput{
+		Bucket: aws.String(bucket),
+		LifecycleConfiguration: &s3.BucketLifecycleConfiguration{
+			Rules: []*s3.LifecycleRule{
+				{
+					// Note: Though as per documentation at https://docs.aws.amazon.com/AmazonS3/latest/API/API_LifecycleRule.html the Filter field is
+					// optional, if not specified the SDK API fails with `Malformed XML` error code. Cross verified same behavior with aws-cli client as well.
+					// Please do not remove it.
+					Filter: &s3.LifecycleRuleFilter{
+						Prefix: aws.String(""),
+					},
+					AbortIncompleteMultipartUpload: &s3.AbortIncompleteMultipartUpload{
+						DaysAfterInitiation: aws.Int64(7),
+					},
+					Status: aws.String(s3.ExpirationStatusEnabled),
+				},
+			},
+		},
+	}
+
+	_, err := c.S3.PutBucketLifecycleConfigurationWithContext(ctx, putBucketLifecycleConfigurationInput)
+	return err
+}
+
+// DeleteBucketIfExists deletes the s3 bucket with name <bucket>. If it does not exist,
+// no error is returned.
+func (c *S3Client) DeleteBucketIfExists(ctx context.Context, bucket string) error {
+	if _, err := c.S3.DeleteBucketWithContext(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)}); err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return nil
+			}
+			if aerr.Code() == "BucketNotEmpty" {
+				if err := c.DeleteObjectsWithPrefix(ctx, bucket, ""); err != nil {
+					return err
+				}
+				return c.DeleteBucketIfExists(ctx, bucket)
+			}
+		}
+		return err
```

Comment:
Let's wrap all errors coming from external packages.
```suggestion
		return errors.Wrap(err, "could not delete bucket")
```